File: apps/ios/Voiid/Voiid/Onboarding/CreateProfileScreen.swift
//
//  CreateProfileScreen.swift
//  Voiid
//
//  Profile setup, step 2 of 2 — the optional half, and the page that actually SAVES.
//
//  Built to the design source (`Voiid Ui/Screens/ProfileExtrasScreen.swift`) through
//  `OnboardingKit`.
//
//  ── EVERYTHING HERE IS SKIPPABLE, AND THE SCREEN SAYS SO ────────────────────────
//  Email and bio are both optional, which is precisely why they were split off step 1: mixed in
//  with the required fields, "optional" is a word in a label that nobody reads. On their own
//  page with a visible Skip, it is a real choice.
//
//  So the primary button is ALWAYS enabled (bar a malformed email). There is nothing to
//  complete — pressing it with both fields empty is a legitimate outcome, identical to Skip. The
//  two exist together because they mean different things to the user ("I'm done" vs "not now"),
//  not because they do different things to the data.
//
//  ── THE SAVE HAPPENS ONCE, HERE ─────────────────────────────────────────────────
//  Step 1 collects identity and hands over a `ProfileDraft` without writing anything, so a user
//  who backs out has not half-created an account. Both Continue and Skip call `submit()`, which
//  is the single `ProfileService.updateProfile` for the whole flow. Skip simply submits a draft
//  with the optional fields empty — it is not a way to avoid saving the required ones.
//
//  ── EMAIL IS VALIDATED ONLY IF TYPED ────────────────────────────────────────────
//  The account is keyed on a verified phone number, so email is a recovery convenience rather
//  than an identity. An empty field passes. A field with something in it has to look like an
//  address — catching a typo here is worth it precisely BECAUSE the address is for recovery,
//  which is the one moment a wrong address cannot be corrected.
//

import SwiftUI

struct CreateProfileScreen: View {
    /// Everything collected on step 1.
    let draft: ProfileDraft
    let onFinish: () -> Void

    @EnvironmentObject var session: AppSession

    @State private var email = ""
    @State private var bio = ""
    @State private var saving = false
    @State private var errorText: String?

    /// Empty is fine — see the header. A typed value has to look like an address.
    private var emailValid: Bool {
        let e = email.trimmingCharacters(in: .whitespaces)
        if e.isEmpty { return true }
        // Deliberately loose. Strict RFC-5322 matching rejects valid addresses, and the only
        // real test is a confirmation mail.
        let parts = e.split(separator: "@")
        return parts.count == 2 && !parts[0].isEmpty && parts[1].contains(".")
            && !parts[1].hasPrefix(".") && !parts[1].hasSuffix(".")
    }

    var body: some View {
        ZStack {
            VoiidBrand.ground.ignoresSafeArea()

            ScrollView {
                VStack(spacing: 0) {
                    OnboardingHeader(
                        title: .stacked("A few more", accent: "details"),
                        blurb: "Both are optional — you can add them later."
                    )

                    VStack(spacing: 12) {
                        OnboardingField(
                            icon: "envelope",
                            label: "Email address",
                            prompt: "Enter your email address",
                            text: $email,
                            keyboard: .emailAddress,
                            contentType: .emailAddress,
                            autocapitalization: .never
                        ) {
                            if !email.isEmpty && !emailValid {
                                Image(systemName: "exclamationmark.circle")
                                    .font(.system(size: 20, weight: .medium))
                                    .foregroundColor(VoiidColor.warning)
                            }
                        }

                        Text("Used to help you recover your account.")
                            .font(VoiidFont.rounded(12.5))
                            .foregroundColor(VoiidBrand.textDim)
                            .frame(maxWidth: .infinity, alignment: .leading)
                            .padding(.leading, 4)

                        OnboardingField(
                            icon: "pencil.line",
                            label: "Bio (optional)",
                            prompt: "Tell the world about yourself",
                            text: $bio,
                            characterLimit: 120
                        )
                    }
                    .padding(.top, VoiidSpacing.lg)

                    if let errorText {
                        Text(errorText)
                            .font(VoiidFont.rounded(13))
                            .foregroundColor(VoiidColor.error)
                            .multilineTextAlignment(.center)
                            .fixedSize(horizontal: false, vertical: true)
                            .padding(.top, VoiidSpacing.sm)
                    }

                    privacyNote
                        .padding(.top, VoiidSpacing.lg)
                }
                .padding(.horizontal, VoiidSpacing.lg)
                .padding(.bottom, 210)
            }
            .scrollIndicators(.hidden)
            .scrollDismissesKeyboard(.interactively)

            VStack(spacing: 0) {
                Spacer(minLength: 0)
                footer
            }
            .ignoresSafeArea(edges: .bottom)
        }
        .preferredColorScheme(.dark)
        .navigationBarTitleDisplayMode(.inline)
        .toolbarBackground(.hidden, for: .navigationBar)
    }

    // MARK: Privacy note

    private var privacyNote: some View {
        HStack(spacing: VoiidSpacing.md) {
            Circle()
                .fill(VoiidBrand.lime.opacity(0.10))
                .frame(width: 46, height: 46)
                .overlay {
                    Image(systemName: "lock.shield")
                        .font(.system(size: 19, weight: .medium))
                        .foregroundColor(VoiidBrand.lime)
                }

            VStack(alignment: .leading, spacing: 3) {
                Text("You control what you share")
                    .font(VoiidFont.rounded(15, .semibold))
                    .foregroundColor(VoiidBrand.text)
                Text("Your email is never shown to other people, and you can change any of this in Settings.")
                    .font(VoiidFont.rounded(12.5))
                    .foregroundColor(VoiidBrand.textDim)
                    .fixedSize(horizontal: false, vertical: true)
            }

            Spacer(minLength: 0)
        }
        .padding(.horizontal, VoiidSpacing.md)
        .padding(.vertical, 14)
        .background(VoiidBrand.card)
        .clipShape(RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: VoiidRadius.lg, style: .continuous)
                .stroke(VoiidBrand.hairline, lineWidth: 1)
        )
        .accessibilityElement(children: .combine)
    }

    // MARK: Footer

    private var footer: some View {
        OnboardingFooter {
            OnboardingKitButton(title: saving ? "Saving…" : "Finish",
                                enabled: emailValid && !saving) {
                submit(includeExtras: true)
            }

            // Skip saves the SAME required fields — it only declines the optional ones. It is
            // not a way out of creating the profile, and it must not look like one.
            Button("Skip for now") {
                Haptics.tap()
                submit(includeExtras: false)
            }
            .font(VoiidFont.rounded(15))
            .foregroundColor(VoiidBrand.textDim)
            .buttonStyle(PressableButtonStyle())
            .disabled(saving)

            StepDots(current: 1, total: 2)
        }
    }

    // MARK: Save

    /// The single write for the whole profile flow.
    private func submit(includeExtras: Bool) {
        guard !saving else { return }
        saving = true
        errorText = nil

        let cleanEmail = includeExtras ? email.trimmingCharacters(in: .whitespaces) : ""
        let cleanBio = includeExtras ? bio.trimmingCharacters(in: .whitespaces) : ""

        Task {
            do {
                _ = try await ProfileService.shared.updateProfile(
                    fullName: draft.fullName,
                    email: cleanEmail,
                    bio: cleanBio.isEmpty ? nil : cleanBio,
                    username: draft.username
                )
                session.profile.fullName = draft.fullName
                session.profile.email = cleanEmail
                session.profile.bio = cleanBio
                Haptics.success()
                onFinish()
            } catch let APIError.http(status, _, _) where status == 409 {
                // The name was free on step 1 and was taken in between. Sending the user back is
                // the only honest fix — the field that has to change is on the other page.
                errorText = "That username was just taken. Go back and choose another."
                Haptics.error()
            } catch {
                errorText = (error as? APIError)?.errorDescription ?? "Couldn't save profile."
                Haptics.error()
            }
            saving = false
        }
    }
}

File: apps/android/app/src/main/java/com/voiid/app/onboarding/CreateProfileScreen.kt
package com.voiid.app.onboarding

import androidx.compose.material.icons.outlined.Edit
import androidx.compose.material.icons.outlined.ErrorOutline
import androidx.compose.material.icons.outlined.MailOutline
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import com.voiid.app.ui.theme.VoiidSpacing
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Cancel
import androidx.compose.material.icons.outlined.GppGood
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import kotlinx.coroutines.launch
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.voiid.app.main.MediaCache
import com.voiid.app.model.AppSession
import com.voiid.app.net.MediaService
import com.voiid.app.net.ProfileService
import com.voiid.app.net.TokenStore
import com.voiid.app.ui.components.LocalVoiidHaptics
import com.voiid.app.ui.theme.VoiidColor
import com.voiid.app.ui.theme.VoiidFont

/**
 * Onboarding OPTIONAL details — email + bio, port of iOS `CreateProfileScreen`.
 *
 * Everything REQUIRED was collected on the previous screen; both fields here are optional and
 * "Skip for now" saves the SAME required fields while declining only these extras. It is not a
 * way out of creating the profile, and must not look like one.
 *
 * This is also THE SINGLE SERVER WRITE for the whole flow: name, username, photo (if chosen),
 * then email/bio when provided.
 */
@Composable
fun CreateProfileScreen(
    session: AppSession,
    draft: SignupDraft,
    onBack: () -> Unit,
    onFinish: () -> Unit,
) {
    val haptics = LocalVoiidHaptics.current
    val context = LocalContext.current
    val scope = androidx.compose.runtime.rememberCoroutineScope()

    var email by remember { mutableStateOf("") }
    var bio by remember { mutableStateOf("") }
    var saving by remember { mutableStateOf(false) }
    var errorText by remember { mutableStateOf<String?>(null) }

    // Empty is fine. A typed value has to LOOK like an address — deliberately loose, because
    // strict matching rejects valid addresses and the real test is a confirmation mail. iOS.
    val emailValid = email.trim().isEmpty() || run {
        val parts = email.trim().split("@")
        parts.size == 2 && parts[0].isNotEmpty() && parts[1].contains(".") &&
            !parts[1].startsWith(".") && !parts[1].endsWith(".")
    }
    val canFinish = emailValid && !saving

    fun submit(includeExtras: Boolean) {
        if (saving) return
        saving = true
        errorText = null
        scope.launch {
            try {
                // Photo first, so the profile write can reference its key.
                var photoKey: String? = null
                draft.photoJpeg?.let { bytes ->
                    photoKey = MediaService(TokenStore.get(context)).uploadProfilePhoto(bytes)
                    // Local-first: cache under the key so the avatar shows instantly everywhere.
                    MediaCache.putData(context, photoKey!!, bytes)
                }
                ProfileService(context).updateProfile(
                    fullName = draft.fullName,
                    username = draft.username,
                    photoUrl = photoKey,
                    email = if (includeExtras && email.isNotBlank()) email.trim() else null,
                    bio = if (includeExtras && bio.isNotBlank()) bio.trim() else null,
                )
                session.profile = session.profile.copy(
                    fullName = draft.fullName,
                    username = draft.username,
                    email = if (includeExtras) email.trim() else session.profile.email,
                    bio = if (includeExtras) bio.trim() else session.profile.bio,
                    photoURL = photoKey ?: session.profile.photoURL,
                )
                haptics.success()
                onFinish()
            } catch (e: com.voiid.app.net.ApiError.Http) {
                if (e.status == 409) {
                    // The name was free a moment ago and was taken in between. Sending the user
                    // back is the only honest fix — the field that has to change is on the other
                    // page. Mirrors iOS.
                    errorText = "That username was just taken. Go back and choose another."
                } else {
                    errorText = e.message ?: "Couldn't save profile."
                }
                haptics.error()
            } catch (e: Exception) {
                errorText = e.message ?: "Couldn't save profile."
                haptics.error()
            }
            saving = false
        }
    }

    OnboardingScaffold(
        footer = {
            OnboardingKitButton(
                title = if (saving) "Saving…" else "Finish",
                enabled = canFinish,
                busy = saving,
            ) { submit(includeExtras = true) }

            Text(
                "Skip for now",
                style = VoiidFont.rounded(15),
                color = VoiidBrand.textDim.copy(alpha = if (saving) 0.5f else 1f),
                modifier = Modifier
                    .align(Alignment.CenterHorizontally)
                    .clickable(
                        interactionSource = remember { MutableInteractionSource() },
                        indication = null,
                        enabled = !saving,
                    ) { haptics.tap(); submit(includeExtras = false) },
            )

            StepDots(current = 1, total = 2)
        },
    ) {
        Spacer(Modifier.height(VoiidSpacing.md))

        OnboardingHeader(
            title = OnboardingTitleSpec.Stacked("A few more", "details"),
            blurb = "Both are optional — you can add them later.",
        )

        Column(
            Modifier.fillMaxWidth().padding(top = VoiidSpacing.lg),
            verticalArrangement = Arrangement.spacedBy(12.dp),
        ) {
            OnboardingField(
                icon = Icons.Outlined.MailOutline,
                label = "Email address",
                prompt = "Enter your email address",
                value = email,
                onValueChange = { email = it },
                keyboardType = KeyboardType.Email,
                capitalization = KeyboardCapitalization.None,
                trailing = {
                    // WARNING, not error, and no red border: an address still being typed is
                    // incomplete rather than wrong.
                    if (email.isNotEmpty() && !emailValid) {
                        Icon(Icons.Outlined.ErrorOutline, null, tint = VoiidColor.warning,
                            modifier = Modifier.size(20.dp))
                    }
                },
            )

            Text(
                "Used to help you recover your account.",
                style = VoiidFont.rounded(12.5f), color = VoiidBrand.textDim,
                modifier = Modifier.fillMaxWidth().padding(start = 4.dp),
            )

            // The counter lives INSIDE the field, bottom-right — `characterLimit` draws it.
            OnboardingField(
                icon = Icons.Outlined.Edit,
                label = "Bio (optional)",
                prompt = "Tell the world about yourself",
                value = bio,
                onValueChange = { bio = it },
                characterLimit = 120,
                imeAction = ImeAction.Done,
            )
        }

        errorText?.let {
            Text(it, style = VoiidFont.rounded(13), color = VoiidColor.error,
                textAlign = androidx.compose.ui.text.style.TextAlign.Center,
                modifier = Modifier.fillMaxWidth().padding(top = VoiidSpacing.sm))
        }

        PrivacyNote(modifier = Modifier.padding(top = VoiidSpacing.lg))

        Spacer(Modifier.height(VoiidSpacing.xl))
    }
}

/** The one reassurance on the screen: nothing here is public. */
@Composable
private fun PrivacyNote(modifier: Modifier = Modifier) {
    Row(
        modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(16.dp))
            .background(VoiidColor.fieldFill)
            .border(1.dp, VoiidColor.fieldBorder, RoundedCornerShape(16.dp))
            .padding(14.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            Modifier.size(46.dp).clip(CircleShape)
                .background(VoiidColor.primary.copy(alpha = 0.10f)),
            contentAlignment = Alignment.Center,
        ) {
            Icon(Icons.Outlined.GppGood, null, tint = VoiidColor.primary, modifier = Modifier.size(22.dp))
        }
        Column(Modifier.weight(1f)) {
            Text("You control what you share",
                 style = VoiidFont.rounded(15, FontWeight.SemiBold), color = VoiidColor.textPrimary)
            Text(
                "Your email is never shown to other people, and you can change any of this in Settings.",
                style = VoiidFont.rounded(12), color = VoiidColor.textSecondary,
                modifier = Modifier.padding(top = 2.dp),
            )
        }
    }
}
